src/components/products/product-list/ko/runtime/product-list.ts

```diff
@@ -28,32 +37,86 @@ export class ProductList {
         this.products = ko.observableArray();
         this.selectedProductId = ko.observable();
         this.working = ko.observable(true);
+
+        this.pattern = ko.observable();
+        this.page = ko.observable(1);
+        this.hasPrevPage = ko.observable(false);
+        this.hasNextPage = ko.observable(false);
+        this.hasPager = ko.computed(() => this.hasPrevPage() || this.hasNextPage());
     }
 
     @OnMounted()
     public async initialize(): Promise<void> {
         await this.usersService.ensureSignedIn();
-        await this.loadProducts();
+        await this.searchByName();
 
         this.router.addRouteChangeListener(this.onRouteChange);
     }
 
-    private async loadProducts(): Promise<void> {
+    private getProductId(): string {
+        const route = this.router.getCurrentRoute();
+        const queryParams = new URLSearchParams(route.hash);
+        const productId = queryParams.get("productId");
+
+        return productId;
+    }
+
+    private onRouteChange(): void {
+        const productId = this.getProductId();
+
+        if (!productId) {
+            return;
+        }
+
+        this.selectedProductId(`/products/${productId}`);
+    }
+
+    public getProductUrl(product: Product): string {
+        return product.id.replace("/products/", "/product#?productId=");
+    }
+
+    public dispose(): void {
+        this.router.removeRouteChangeListener(this.onRouteChange);
+    }
+
+    public async onEnter(data?: any, event?: KeyboardEvent): Promise<boolean> {
+        if(event && event.keyCode === 13) {
+            await this.searchByName();
+        }
+        return true;
+    }
+
+    public async searchByName(): Promise<void> {
+        const currentPattern = this.pattern();
+        if (this.lastPattern !== currentPattern) {
+            this.page(1);
+        }
+
+        const pageNumber = this.page() - 1;
+        const query: PatternFilter = {
+            pattern: this.pattern(),
+            skip: pageNumber * Constants.defaultPageSize,
+            take: Constants.defaultPageSize
+        };
+
+        this.working(true);
+
         try {
-            this.working(true);
-            const products = await this.productService.getProducts();
+            const itemsPage = await this.productService.getProductsPage(query);
+            const contracts = itemsPage.value;
 
-            this.products(products);
+            this.hasPrevPage(pageNumber > 0);
+            this.hasNextPage(!!itemsPage.nextLink);
 
-            let productId = this.getProductId();
+            const searchItems = [];
 
-            if (productId) {
-                this.selectedProductId(`/products/${productId}`);
-            }
-            else if (products.length > 0) {
-                productId = products[0].id;
-                this.selectProduct(productId);
+            if (contracts) {
+                contracts.map(item => searchItems.push(new Product(item)));            
```

Comment:
I would move this inside service method, so that it returned models rather than contracts:
`contracts.map(item => searchItems.push(new Product(item)));`